### ISSUE TITLE
代码小问题求助

### ISSUE THREAD

**Issue by chenbinghui1** [layumi/Person_reID_baseline_pytorch#92]
在IDE模型中，遇到了一个bug不知道怎么回事，当前提供的模型代码结构是这样，resnet50+classifier（其中包括add_block和classifier for 751 classes），这是在model.py中定义的，但是我把其中的classier for751 部分拿到外面定义，也就是最后的分类层在train.,py中定义，即加了linear=nn.Sequential(nn.linear(512,751).cuda())，然后 outputs=model（inputs）变为outputs=model(inputs), outputs=linear(outputs)后训练，训练损失还有acc都和之前分类层放在model里面定义是一样的 但是 test的时候R@1和mAP分别下降了10和20个点，请问这是怎么回事，刚用pytorch不知道为什么，困扰了两天了

**Comment by layumi** [comment]
Could you provide your model definition code?

**Comment by chenbinghui1** [comment]
in model.py :
class ClassBlock(nn.Module):
    def __init__(self, input_dim, class_num, dropout=True, relu=True, num_bottleneck=512):
        super(ClassBlock, self).__init__()
        add_block = []
        add_block += [nn.Linear(input_dim, num_bottleneck)] 
        add_block += [nn.BatchNorm1d(num_bottleneck)]
        if relu:
            add_block += [nn.LeakyReLU(0.1)]
        if dropout:
            add_block += [nn.Dropout(p=0.5)]
        add_block = nn.Sequential(*add_block)
        add_block.apply(weights_init_kaiming)
        self.add_block = add_block

        ## comment the original code ##
        #classifier = []
        #classifier += [nn.Linear(num_bottleneck, class_num)]
        #classifier = nn.Sequential(*classifier)
        #classifier.apply(weights_init_classifier)
        #self.classifier = classifier
        
    def forward(self, x):
        x = self.add_block(x)
        #x = self.classifier(x)  ## comment the original code ##
        return x




In train.py:
........
linear = nn.Sequential(nn.Linear(feature_dim, class_num)).cuda() ## added code ##
def train_model(model, criterion, optimizer, scheduler, num_epochs=25):
........
for data in dataloaders[phase]:
                # get the inputs
                inputs, labels = data
                now_batch_size,c,h,w = inputs.shape
                if now_batch_size<opt.batchsize: # skip the last batch
                    continue
                #print(inputs.shape)
                # wrap them in Variable
                if use_gpu:
                    inputs = Variable(inputs.cuda())
                    labels = Variable(labels.cuda())
                else:
                    inputs, labels = Variable(inputs), Variable(labels)

                # zero the parameter gradients
                optimizer.zero_grad()
                # forward
                outputs = model(inputs)
                outputs = linear(outputs) ## added code ##
                .......



Add some new codes and comment some old codes, and the rest part of codes are not changed.

**Comment by layumi** [comment]
Hi @chenbinghui1 ,
1. You do not initial the linear layer that you added?
2. Have you add the parameter to the optimizer? So the sgd optimizer may ignore the parameter update.

**Comment by chenbinghui1** [comment]
@layumi 
1.I have tested the initial method that does effect much
2.thank you for remind, i just observe this issue and am training by setting a small learning rate.
